### PR TITLE
[第九組] Refactor/error handling

### PR DESCRIPTION
**Pull request by Steven0811**
## Purpose
1. Add error handler allowing all errors being catched by Sentry.
2. Add logger to record errors and users' activities.

## Change
1. Add ErrorHandler:
Track errors such as HTTPException, ValueError, AttributeError and DomainMismatchError. After ErrorHandler receives an error, it will transmit the error to Sentry and the error will be marked as handled. At the same time, the error will be recorded by logger. Notice that if the status code of the exception is 400 , which means it is client error, the logger level is WARNING. If it is not the above situation, the logger level is ERROR.
2.  Add logger:
The logger will record all exceptions and main activities of users such as login, search news, register, etc. The logger will put logs on streamer, app.log and app_rotating.log while app.log shows all activities that logger level is greater than INFO. On the other hand, app_rotating.log only shows exception. Following is a simple logs example.
```
2024-12-19 05:26:39,494 - src.auth.service - check_user_password_is_correct - INFO - User Tuna authenticated successfully. 
2024-12-19 05:26:39,495 - src.auth.service - create_access_token - DEBUG - Access token created successfully. 
2024-12-19 05:26:39,496 - src.users.router - login_for_access_token - INFO - User Tuna logged in successfully. 
2024-12-19 05:26:39,515 - src.users.router - read_users_me - INFO - User Tuna accessed their profile 
2024-12-19 05:26:39,537 - src.prices.router - get_necessities_prices - INFO - Fetching prices for category 'None' and commodity 'None'. 
2024-12-19 05:26:45,317 - src.news.router - read_user_news - INFO - User news data retrieved successfully. 
2024-12-19 05:26:58,461 - src.news.service - toggle_upvote - INFO - Upvote removed. 
2024-12-19 05:26:59,519 - src.news.service - toggle_upvote - INFO - Article upvoted. 
2024-12-19 05:27:04,808 - src.prices.router - get_necessities_prices - INFO - Fetching prices for category 'None' and commodity 'None'. 
2024-12-19 05:27:15,254 - src.users.router - read_users_me - INFO - User Tuna accessed their profile 
2024-12-19 05:27:29,616 - ErrorHandler - catch_error - WARNING - Client error in /usr/local/lib/python3.10/site-packages/fastapi/security/oauth2.py at line 478: 401: Not authenticated 
2024-12-19 05:29:25,534 - ErrorHandler - catch_error - WARNING - Client error in /app/src/news/router.py at line 141: 400: Invalid model specified. 

```

### CODE REVIEW

**Comment by seanpai96** [comment]
Please fix your conflict so I can merge!

File: backend/src/auth/service.py [hunk 1]
```diff
@@ -37,12 +47,17 @@ def create_access_token(data, expires_delta=None):
                           If not provided, the token will expire in 15 minutes.
     :return: A JWT access token as a string.
     """
+    logger = Logger(__name__, "create_access_token").get_logger()
     to_encode = data.copy()
     if expires_delta:
         expire = datetime.utcnow() + expires_delta
     else:
-        expire = datetime.utcnow() + timedelta(minutes=Auth.DEFAULT_ACCESS_TOKEN_EXPIRE_MINUTES)
+        expire = datetime.utcnow() + timedelta(minutes=15)
```
Comment (samlu00412) [inline backend/src/auth/service.py:55]: 我覺得這有點像magical number，原本的寫法好像好一點

File: backend/src/crawler/udn_crawler.py [hunk 2]
```diff
@@ -83,63 +90,81 @@ def _fetch_news(self, page: int, search_term: str) -> list[Headline]:
...
-            "channelId": self.CHANNEL_ID,
-            "type": "searchword",
-        }
+        try:
+            return {
+                "page": page,
+                "id": f"search:{search_term}",
+                "channelId": self.CHANNEL_ID,
+                "type": "searchword",
+            }
```
Comment (samlu00412) [inline backend/src/crawler/udn_crawler.py:99]: 這四個屬性看起來是一篇新聞或文章的基本資訊，建議可以實作一個class儲存這些datas

**Review by samlu00412 — COMMENTED** [review]
整體上來說是一個很好的PR，將logger跟各種error_handling整合在各自的檔案中

**Comment by haruuu1214** [comment]
### 檔案修改

`backend/src/news/router.py`:
  - 移除了 `status` import，但新增了 `SQLAlchemyError` import，這樣的修改應該是基於實際需要的。
  - 新增了 `Logger`，並在多處使用，這有助於提供更好的日誌記錄。
  - 增加了 try-except 區塊來處理 SQLAlchemy 和 AttributeError 的例外情況，這有助於提升 API 的穩定性。
  - 為 `news_summary_custom_model` 添加了一個 logger，並在生成摘要時加入了異常處理，這樣的改動使程式碼更加穩健。
  - 建議在 logging 信息中包括更多上下文資料，以便更好地追蹤問題。

`backend/src/prices/router.py`:
  - 新增了 `HTTPException` 和 `Logger`，並在 `get_necessities_prices` 中增加了日誌記錄和例外處理，這有助於更好地調試和錯誤處理。
  - 建議在 logging 信息中包括 API 的回應狀態和內容，以便更好地分析問題。

`backend/src/users/router.py`:
  - 新增了 `HTTPException` 和 `Logger`，並在多處增加了日誌記錄和例外處理，這有助於提升程式碼的可維護性和可調試性。
  - 在 `create_user` 方法中添加了用戶名重複的檢查，這有助於防止資料庫中出現重複資料。
  - 建議在 logging 信息中包括更多上下文資料，例如用戶操作的詳細資訊，以便更好地追蹤問題。

### 總結
- 這些修改整體上增加了程式碼的穩定性和可維護性，特別是通過增加日誌記錄和例外處理。
- 建議在 logging 信息中包括更多上下文資料，以便更好地追蹤和解決問題。
- 確認新增檔案的用途和正確性，以確保他們符合項目的需求。

整體改動益無窮，
穩定維護愈久長。
確認檔案用途對，
步步精進計劃強！

**Comment by patrickhicetc** [comment]
有些try except包的程式碼太多，但總體來說該處理的錯誤都有抓到，logger也都有紀錄，是個優秀的PR

**Comment by Rogue8464** [comment]
這份PR加入了錯誤處理和日誌記錄以提供程式運行的穩定性，模組化清晰。建議精細化 try-except 區塊，並在日誌中增加更多上下文資訊以便問題追蹤。

**Comment by mlcp29th-hi** [comment]
The recent updates to the codebase have introduced increased error handling in various components. This is a crucial aspect of improving the robustness and reliability of the application. However, I would like to suggest some additional improvements and best practices for error handling to further ensure that we are capturing edge cases and providing meaningful feedback to consumers of our API.

#### Suggested Improvements

1. **More Specific Exception Handling**:
   - It's great to see that `HTTPException` is being raised in some parts when encountering errors. However, consider more granular exception handling for specific operations, especially in database interactions and external requests (e.g., HTTP calls). This will provide more context to the consumer about where the error originated from.

   ```python
   except SQLAlchemyError as db_err:
       logger.error(f"Database error: {str(db_err)}")
       raise HTTPException(status_code=500, detail="Database operation failed.")
   ```

2. **Consistent Logging**:
   - While many methods utilize logging, ensure that all exceptions are logged consistently. This includes not just successful operations but especially errors, capturing useful context information like input parameters or the state of the application at the time of the error.

3. **Specific Error Responses**:
   - Instead of returning a generic message (e.g., "An unexpected error occurred"), provide more specific error messages tailored to the type of operation being performed. This can help with debugging and enhance the user experience.

   ```python
   except ValueError as ve:
       logger.warning(f"Value error: {str(ve)}")
       raise HTTPException(status_code=400, detail=f"Invalid input: {str(ve)}")
   ```

4. **Avoiding Broad Exception Catch**:
   - In several instances, broad `except Exception as e:` is used. It's generally better to catch more specific exceptions unless there’s a clear intention to catch all possible exceptions. This practice helps in avoiding the masking of bugs.

5. **Structured Error Handling Middleware**:
   - Consider implementing centralized error handling middleware that can catch all unhandled exceptions throughout the application. This could greatly reduce code duplication and ensure uniformity in how errors are presented.

   ```python
   @app.middleware("http")
   async def error_handling_middleware(request: Request, call_next):
       try:
           return await call_next(request)
       except HTTPException as exc:
           return JSONResponse(status_code=exc.status_code, content={"message": exc.detail})
       except Exception as e:
           logger.error(f"Unhandled error: {str(e)}")
           return JSONResponse(status_code=500, content={"message": "Internal server error"})
   ```

6. **Unit Tests for Error Scenarios**:
   - Ensure that the error handling pathways are covered in your test cases. Create tests that simulate failure conditions for database calls, network requests, and invalid input to confirm that the error responses are accurate and logged.

7. **Handling JSON Decoding Errors**:
   - Since the application interfaces with external APIs, adding error handling specifically for JSON decoding issues (like the `JSONDecodeError` you already have) should be considered at all places where the response is parsed to avoid runtime errors.

8. **Graceful Failures**:
   - When facing failures, inform users gracefully about what went wrong and possibly how they can go about correcting it, instead of a generic message.

#### Conclusion
Overall, the error handling in this pull request demonstrates a commitment to robustness. With the suggested improvements, we can further enhance the reliability and maintainability of the code. Having a well-structured approach to error management will not only assist developers but will also improve the user experience by providing more informative feedback for troubleshooting.

**Comment by Yuannzh1** [comment]
#### Try-Except Block 評估
1. **`dependencies.py` 中的 `authenticate_user_token` 函數**：
   - Try-Except 內部處理的事情：JWT 解碼、從 payload 獲取 username、查詢用戶資料。
   - 這個 Try-Except 塊內的步驟相對簡單，每個步驟都與異常處理直接相關，處理適當。

2. **`service.py` 中的 `verify` 函數**：
   - Try-Except 內部處理的事情：密碼驗證。
   - 這個 Try-Except 塊內部步驟單一，異常處理合理。

3. **`service.py` 中的 `check_user_password_is_correct` 函數**：
   - Try-Except 內部處理的事情：查詢用戶、驗證密碼。
   - 這個 Try-Except 塊內部的步驟適當，每個步驟都與異常處理直接相關。

4. **`service.py` 中的 `create_access_token` 函數**：
   - Try-Except 內部處理的事情：JWT 編碼。
   - 這個 Try-Except 塊內的步驟單一，異常處理合理。

### 建議
1. **Try-Except 塊分解**：
   - 雖然目前的 Try-Except 塊大部分都很合理，但可以進一步分解一些較大的塊，以提高可讀性和可維護性。例如，在 `authenticate_user_token` 函數內，可以將 JWT 解碼和查詢用戶分成兩個 Try-Except 塊。

總的來說，這個 PR 在異常處理和日誌記錄方面做得很好，但可以通過進一步分解 Try-Except 塊和提供更具體的日誌消息來進一步改進。

**Comment by percykugood** [comment]
建議可以在try-except裡面包含少一點的東西，除此之外沒有太大的問題，都有抓到該抓到的錯誤，是個很好的PR，果然是鮪魚大神

《鮪影隨PR行》

鮪魚游動海中深，
波濤洶湧展雄心。
PR如潮逐浪來，
代碼如鮪吞無痕。

File: backend/src/auth/service.py (hunk 1 above)
Comment (WANG-YongLE) [inline backend/src/auth/service.py:55]: 你改他幹啥阿

File: backend/src/crawler/udn_crawler.py [hunk 3]
```diff
@@ -64,17 +66,22 @@ def get_headline(
         # If 'page' is a tuple, unpack it and create a range representing those pages (inclusive).
         # If 'page' is an int, create a list containing only that single page number.
         # page_range = range(*page) if isinstance(page, tuple) else [page]
-        if isinstance(page, tuple):
-            start_page, end_page = page
-            page_range = range(start_page, end_page + 1)
-        else:
-            page_range = [page]
+        logger = Logger(__name__, "get_headline").get_logger()
```
Comment (WANG-YongLE) [inline backend/src/crawler/udn_crawler.py:69]: 你把logger分好細

**Comment by WANG-YongLE** [comment]
我個人覺得沒啥問題，非常優秀

File: backend/src/auth/service.py (hunk 1 above)
Comment (dee9c-git) [inline backend/src/auth/service.py:55]: ~~datetime.utcnow()~~
https://blog.miguelgrinberg.com/post/it-s-time-for-a-change-datetime-utcnow-is-now-deprecated
Use this (I think) instead
```python
datetime.now(timezone.utc)
```
also magic number?

File: backend/src/news/service.py [hunk 4]
```diff
@@ -99,25 +107,33 @@ def toggle_upvote(article_id, user_id, news_db):
...
+            logger.info("Upvote removed.")
+            return "Upvote removed"
+        else:
+            insert_statement = insert(user_news_association_table).values(
+                news_articles_id=article_id, user_id=user_id
+            )
+            news_db.execute(insert_statement)
+            news_db.commit()
+            logger.info("Article upvoted.")
+            return "Article upvoted"
```
Comment (dee9c-git) [inline backend/src/news/service.py:135]: Large ```try``` statement

File: backend/tests/integration/test_news_endpoint.py [hunk 5]
```diff
@@ -151,20 +151,6 @@ def test_search_news(mocker):
...
-    mock_openai(mocker, openai_response)
-
-    request_body = NewsSumaryRequestSchema(content="Test news content")
-    response = client.post("/api/v1/news/news_summary", json=request_body.dict(), headers=headers)
-
-    assert response.status_code == 200
-    json_response = response.json()
-    assert json_response["summary"] == "test impact"
-    assert json_response["reason"] == "test reason"
-
```
Comment (dee9c-git) [inline backend/tests/integration/test_news_endpoint.py:167]: Um, 
~~is this because you couldn't get news_summary() to work?~~

**Review by New9c — COMMENTED** [review]
Pretty good all things considered.
~~Not sure why you deleted test_news_summary() though~~

**Comment by hamster0717** [comment]
### **Strengths**

**Enhanced Error Handling:**
   - Introduction of specific exceptions (e.g., `HTTPException`) and use of HTTP status codes improves error granularity.
   - Consolidated error handling logic in classes such as `ErrorHandler` and `HTTPExceptionHandler` ensures a consistent approach to managing exceptions.

**Improved Logging Infrastructure:**
   - Custom `Logger` class improves modularity and provides consistent logging formats across modules.
   - Inclusion of log messages for critical operations (e.g., user authentication, token generation, and database actions) enhances system observability.

**Pull request merged**